dipy/reconst/mapmri.py

```diff
@@ -0,0 +1,666 @@
+
+
+def mapmri_psi_matrix(radial_order, mu, rgrad):
+    r"""Compute the MAPMRI psi matrix for the propagator [1]_
+
+
+    Parameters
+    ----------
+    radial_order : unsigned int,
+        an even integer that represent the order of the basis
+    mu: array, shape (3,)
+        scale factors of the basis for x, y, z
+    rgrad: array, shape (N,3)
+        points in the r-space in which evaluate the EAP
+
+
+    References
+    ----------
+    .. [1] Ozarslan E. et. al, "Mean apparent propagator (MAP) MRI: A novel
+    diffusion imaging method for mapping tissue microstructure",
+    NeuroImage, 2013.
+    """
+
+    ind_mat = mapmri_index_matrix(radial_order)
+    n_elem = ind_mat.shape[0]
+    n_rgrad = rgrad.shape[0]
+    K = np.zeros((n_rgrad, n_elem))
+    for j in range(n_elem):
+        K[:, j] = mapmri_psi_3d(ind_mat[j], rgrad, mu)
+
+    return K
+
+
+def mapmri_odf_matrix(radial_order, mu, s, vertices):
+    r"""Compute the MAPMRI ODF matrix [1]_  Eq. 33.
+
+    Parameters
+    ----------
+    radial_order : unsigned int,
+        an even integer that represent the order of the basis
+    mu: array, shape (3,)
+        scale factors of the basis for x, y, z
+    s: unsigned int
+        radial moment of the ODF
+    vertices: array, shape (N,3)
+        points of the sphere shell in the r-space in which evaluate the ODF
+
+
+    References
+    ----------
+    .. [1] Ozarslan E. et. al, "Mean apparent propagator (MAP) MRI: A novel
+    diffusion imaging method for mapping tissue microstructure",
+    NeuroImage, 2013.
+    """
+
+    ind_mat = mapmri_index_matrix(radial_order)
+    n_vert = vertices.shape[0]
+    n_elem = ind_mat.shape[0]
+    odf_mat = np.zeros((n_vert, n_elem))
+    mux, muy, muz = mu
+    # Eq, 35a
+    rho = 1.0 / np.sqrt((vertices[:, 0] / mux) ** 2 +
+                        (vertices[:, 1] / muy) ** 2 + (vertices[:, 2] / muz) ** 2)
+    # Eq, 35b
+    alpha = 2 * rho * (vertices[:, 0] / mux)
+    # Eq, 35c
+    beta = 2 * rho * (vertices[:, 1] / muy)
+    # Eq, 35d
+    gamma = 2 * rho * (vertices[:, 2] / muz)
+    const = rho ** (3 + s) / np.sqrt(2 ** (2 - s) * np.pi **
+                                     3 * (mux ** 2 * muy ** 2 * muz ** 2))
+    for j in range(n_elem):
+        n1, n2, n3 = ind_mat[j]
+        f = np.sqrt(factorial(n1) * factorial(n2) * factorial(n3))
+        odf_mat[:, j] = const * f * \
+            _odf_cfunc(n1, n2, n3, alpha, beta, gamma, s)
+
+    return odf_mat
+
+
+def _odf_cfunc(n1, n2, n3, a, b, g, s):
+    r"""Compute the MAPMRI ODF function from [1]_  Eq. 34.
+
+    References
+    ----------
+    .. [1] Ozarslan E. et. al, "Mean apparent propagator (MAP) MRI: A novel
+    diffusion imaging method for mapping tissue microstructure",
+    NeuroImage, 2013.
+    """
+
+    f = factorial
+    f2 = factorial2
+    sumc = 0
+    for i in range(0, n1 + 1, 2):
+        for j in range(0, n2 + 1, 2):
+            for k in range(0, n3 + 1, 2):
```

Comment:
It looks very easy to cythonize. If performance is an issue we could have this as a future TODO. Not for this PR.